state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -243,27 +247,59 @@ pub struct CcmSwapAmounts {
 }
 
 #[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo)]
-enum CcmSwapState {
-	PrincipalSwapScheduled {
-		principal_swap_id: SwapId,
-		gas_budget: AssetAmount,
-		other_gas_asset: Option<Asset>,
-	},
-	GasSwapScheduled {
-		gas_swap_id: SwapId,
-		principal_output_amount: AssetAmount,
-	},
+enum GasSwapState {
+	OutputReady { gas_budget: AssetAmount },
+	Scheduled { gas_swap_id: SwapId },
+	ToBeScheduled { gas_budget: AssetAmount, other_gas_asset: Asset },
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo)]
+struct DcaState {
+	scheduled_chunk_swap_id: Option<SwapId>,
+	remaining_input_amount: AssetAmount,
+	remaining_chunks: u32,
+	chunk_interval: u32,
+	accumulated_output_amount: AssetAmount,
+}
+
+impl DcaState {
+	// Create initial DCA state; if no dca parameters provided (for non-DCA swaps),
+	// this creates state equivalent to 1 chunk DCA
+	fn new(input_amount: AssetAmount, params: Option<DcaParameters>) -> DcaState {
+		DcaState {
+			scheduled_chunk_swap_id: None,
+			remaining_input_amount: input_amount,
+			remaining_chunks: params.as_ref().map(|p| p.number_of_chunks).unwrap_or(1),
+			// Chunk interval won't be used for non-DCA swaps but seems nicer to
+			// set a reasonable default than unwrap Option when it is needed:
+			chunk_interval: params.as_ref().map(|p| p.chunk_interval).unwrap_or(SWAP_DELAY_BLOCKS),
+			accumulated_output_amount: 0,
+		}
+	}
+
+	fn prepare_next_chunk(&mut self) -> AssetAmount {
+		let chunk_input_amount = self
+			.remaining_input_amount
+			.checked_div(self.remaining_chunks as u128)
+			.unwrap_or(0);
+		self.remaining_chunks = self.remaining_chunks.saturating_sub(1);
+		self.remaining_input_amount =
+			self.remaining_input_amount.saturating_sub(chunk_input_amount);
+		chunk_input_amount
+	}
 }
 
 #[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo)]
 enum SwapRequestState {
 	Ccm {
-		state: CcmSwapState,
+		gas_swap_state: GasSwapState,
 		ccm_deposit_metadata: CcmDepositMetadata,
 		output_address: ForeignChainAddress,
```

Comment:
> we could easily make a requirement for every swap type - a NetworkFee's destination is the vault for example

I'm not sure it is a good idea. The CCM and Regular swaps let users specify output address and we actually use that value once the swap is executed. While we can think of some values that would seem not too "out of place", they aren't going to be used in the same way. We are not egressing network fees like we do for regular swaps, at least not directly, not in USDC. And even if we were, there is only one value (the vault) that makes sense, and it should be stored separately. Why would we store it individually for each swap and open us to the possiblity of accidentally putting the wrong value there?

I do see the how it might make sense to unify Ccm and Regular seeing that they have some fields in common, but I think we would be generalising prematurely. For example, we discussed how the outcome of a swap should be handled outside of the swapping pallet (and I think this is still the plan, just not the priority right now). Once `output_address` is out, I'm not sure it would be worth refactoring further just because of one duplicated field. I agree we we should keep these ideas in mind for the future though.

Reply:
I was thinking output_address could become an enum, something like:

```rust
enum Destination {
	ExternalAccount(ForeignChainAddress),
	InternalAccount(AccountId),
	Vault(StorageItem), // not sure if/how this would work - might just be an inner enum
}
```

Or something like this. Admittedly not fully thought-through, but I feel like this would give us more flexibility - for example it would allow us to add features like 'swap deposits' almost trivially: it's just a swap request with an InternalAccount as destination.

With the current implementation we would need a new variant SwapDeposit with it's own DCA params, it's own (mostly duplicated) match arm in the process_outcome fn etc.. Anyway, not for this PR..